File: Data_Structures/Python/Data_structures/Queue.py

from sys import maxsize
from utils import get_cndtn

class Queue(object):
    def __init__(self, *args, **kwargs):
        self.size = kwargs['size']
        self.queue = [None]*self.size
        self.quantity = 0
        self.head = 0
        self.tail = -1
    
    
    def __increase(self):
        self.quantity = self.quantity + 1
        self.tail = (self.tail + 1) % self.size
    #    print(f"Quantity: {self.quantity}")
        print(f"Tail: {self.tail}")
    
    
    def __decrease(self):
        self.quantity = self.quantity - 1
        self.head = (self.head + 1) % self.size
    #    print(f"Quantity: {self.quantity}")
        print(f"Head: {self.head}")


    def is_full(self):
        return self.quantity == self.size
    

    def is_empty(self):
        return self.quantity == 0


    def enqueue(self, item):
        if self.is_full():
            return (-maxsize - 1)
        self.__increase()
        self.queue[self.tail] = item 
        return item
    

    def dequeue(self):
        if self.is_empty():
            return (-maxsize - 1)
        item = self.queue[self.head]
        self.queue[self.head] = None
        self.__decrease()
        # print(f"Item dequeued: {item}")
        return item
    

    def head(self):
        return self.queue[self.head]
    

    def tail(self):
        return self.queue[self.tail]


    def __str__(self):
        return str(self.queue)

    
    def __len__(self):
        return len(self.queue)


def main():
    queue = Queue(size=20)
    for i in range(1, 21):
        queue.enqueue(i*10)
    print(queue)
    b, _ = get_cndtn(queue)
    i = 0
    while b :
        print(f"Item dequeued: {queue.dequeue()}")
        print(f"Current head: {queue.head}")
        b, _ = get_cndtn(queue)
        i += 1 
    print(queue) 
    print(f"i: {i}")
    for j in range(1, i): 
        queue.enqueue(j*11)
    print(queue) 

if __name__ == "__main__":
    main()

File: Data_Structures/Java/src/Queue.java
public class Queue<T> {

	private final T MIN_VALUE = (T) Integer.valueOf("-214748364");
	
	private int size, quantity = 0, head = 0, tail = -1;
	private Object[] queue = {};
	
	public <T> Queue(int size) {
		this.size = size;
		this.queue = (T[]) new Object[this.size];
	}

	private void decrease() {
		this.quantity--;
		this.head = ++this.head % this.size;
		System.out.printf("Quantity: %d \n", this.quantity);
		System.out.printf("Head: %d \n", this.head);
	}

	private void increase() {
		this.quantity++;
		this.tail = ++this.tail % this.size;
		System.out.printf("Quantity: %d \n", this.quantity);
		System.out.printf("Tail: %d \n", this.tail);
	}

	public boolean isEmpty() {
		return this.quantity == 0;
	}

	public boolean isFull() {
		return this.quantity == this.size;
	}
	
	public T enqueue(T item) {
		if (this.isFull())
			return MIN_VALUE;
		this.increase();
		this.queue[this.tail] = item;
		return item;
	}

	public T dequeued() {
	if (this.isEmpty())
		return MIN_VALUE;
	T item = (T) this.queue[this.head]; 
	this.queue[this.head] = null;
	this.decrease();
	return item;
	}

	public T head() {
		if (this.isEmpty())
			return MIN_VALUE;
		return (T) this.queue[this.head];
	}

	public T tail() {
		if (this.isEmpty())
			return MIN_VALUE;
		return (T) this.queue[this.tail];
	}

	public void printQueue() {
		for (int i = 0; i < queue.length; i++)
			System.out.printf("%s %d %s", i == 0 ? "[" : "," ,queue[i], i == queue.length - 1 ? "]\n" : "" );
	}
	
	public static void main(String[] args) {
	int limit = 20;
	Queue<Integer> queue = new Queue<Integer>(limit);
		for (int i = 1; i <= limit; i++) 
			System.out.printf("Enqueued: %d \n", queue.enqueue(i*10));
		queue.printQueue();
		System.out.printf("Current head: %d \n", queue.head());
		System.out.printf("Current tail: %d \n", queue.tail());
		int range = limit + 1;     
		int to = (int)(Math.random() * range);
		int from = 0;
		if (to > 0) {
			for (; from < to; from++)
				System.out.printf("Dequeued: %d \n",queue.dequeued());
			queue.printQueue();
			System.out.printf("Current head: %d \n", queue.head());
			System.out.printf("Current tail: %d \n", queue.tail());
			to = from;
			range = from + 1; 
			from = (int)(Math.random() * range);
			if (from < to) {
				while (from++ < to)
					System.out.printf("Enqueued: %d \n",queue.enqueue(from*11));
				queue.printQueue();
				System.out.printf("Current head: %d \n", queue.head());
				System.out.printf("Current tail: %d \n", queue.tail());
			}
		}
		
	}

}
